### PR TITLE
aws: add support for credential_process

### LINKED ISSUE #3621: Support AWS profile credential_process
As best I can tell, currently Fluent Bit only supports AWS profiles that directly specify `aws_access_key_id`, `aws_secret_access_key`, and `aws_session_token` in the ~/.aws/credentials file. https://github.com/fluent/fluent-bit/blob/master/src/aws/flb_aws_credentials_profile.c

Please add support for specifying `credential_process` in the ~/.aws/config file. https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sourcing-external.html

Our use case is we have a single process running that fetches role credentials from IoT Core. https://aws.amazon.com/blogs/security/how-to-eliminate-the-need-for-hardcoded-aws-credentials-in-devices-by-using-the-aws-iot-credentials-provider/ So using a credential process makes the most sense.

It would be best if this also abided by the `AWS_CONFIG_FILE` environment variable if it is set.

### PR DESCRIPTION
**Pull request by rittneje**
Signed-off-by: Jesse Rittner <rittneje@gmail.com>

Adds support for credential_process in the AWS config file. https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sourcing-external.html

This implementation adds support for Linux only.

cc @PettitWesley 

Fixes #3621.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [x] Debug log output from testing the change
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

[fluent-bit-debug.txt](https://github.com/fluent/fluent-bit/files/6749989/fluent-bit-debug.txt)
[fb-valgrind-results.txt](https://github.com/fluent/fluent-bit/files/6736427/fb-valgrind-results.txt)

**Documentation**

- [ ] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

File: tests/lib/acutest/acutest.h [hunk 1]
```diff
@@ -820,6 +820,8 @@ test_dump_(const char* title, const void* addr, size_t size)
 void
 test_abort_(void)
 {
+    fflush(stdout);
```
Comment (rittneje) [inline tests/lib/acutest/acutest.h:823]: @PettitWesley I do not actually intend to merge this change as part of this PR. This is just so I can see the test output if it fails. I have a PR open against acutest to fix it for real. https://github.com/mity/acutest/pull/63

**Comment by PettitWesley** [comment]
@zhonghui12 Can you do the first round of review

**Comment by rittneje** [comment]
@PettitWesley @zhonghui12 FYI I realized I misunderstood how the caching logic was supposed to work. I have pushed a fix for that.

**Comment by edsiper** [comment]
is this ready to go ?

File: tests/lib/acutest/acutest.h (hunk 1 above)
Comment (zhonghui12) [reply tests/lib/acutest/acutest.h:823]: So we might need to delete this part before merging it?

Comment (rittneje) [reply tests/lib/acutest/acutest.h:823]: Yes, my plan was to revert this once any review comments are addressed and everything else is good to go.

**Comment by rittneje** [comment]
@zhonghui12 I think those memory leaks it found were just me forgetting to destroy the provider at the end of the tests. Should be fixed now.

**Comment by zhonghui12** [comment]
Hello, did you use valgrind to test? I assume we have no memory leak with this version of code.

**Comment by rittneje** [comment]
> Hello, did you use valgrind to test? I assume we have no memory leak with this version of code.

Yes, the results are attached in the PR description.

File: tests/lib/acutest/acutest.h (hunk 1 above)
Comment (rittneje) [reply tests/lib/acutest/acutest.h]: I decided to just revert this now. I'll try to manage it locally.

**Comment by edsiper** [comment]
@PettitWesley I am making the release in 2-3 hours, is this ready to go ?

**Comment by PettitWesley** [comment]
@edsiper ack

File: src/aws/flb_aws_credentials_http.c [hunk 2]
```diff
@@ -29,11 +29,10 @@
...
 #include <sys/stat.h>
 
-/* HTTP Credentials Endpoints have a standard set of JSON Keys */
-#define AWS_HTTP_RESPONSE_ACCESS_KEY   "AccessKeyId"
-#define AWS_HTTP_RESPONSE_SECRET_KEY   "SecretAccessKey"
-#define AWS_HTTP_RESPONSE_TOKEN        "Token"
-#define AWS_HTTP_RESPONSE_EXPIRATION   "Expiration"
+#define AWS_CREDENTIAL_RESPONSE_ACCESS_KEY   "AccessKeyId"
+#define AWS_CREDENTIAL_RESPONSE_SECRET_KEY   "SecretAccessKey"
+#define AWS_HTTP_RESPONSE_TOKEN              "Token"
```
Comment (PettitWesley) [inline src/aws/flb_aws_credentials_http.c:34]: Based on the docs, credential process can return a key `SessionToken`

Comment (PettitWesley) [reply src/aws/flb_aws_credentials_http.c:34]: Oh I see you solved this a new argument to the parsing function. That works. I probably would've just added both constants and then had the function check for both of them.

**Review by PettitWesley — APPROVED** [review]

**Comment by PettitWesley** [comment]
@edsiper You can merge this one

**Comment by edsiper** [comment]
thanks

**Comment by PettitWesley** [comment]
@rittneje Can you also add docs for this feature? https://github.com/fluent/fluent-bit-docs

**Comment by rittneje** [comment]
@PettitWesley Where would that go exactly? I can't find any docs describing the AWS credentials chain.

**Comment by PettitWesley** [comment]
@rittneje Good point, we don't have any docs on that today. It'd be nice to have something in each of the doc pages for the AWS plugins. Something listing what we support exactly like:

1. Standard env vars...
2. AWS shared credentials file set with AWS_SHARED_CREDENTIALS_FILE. credentials_process is supported...
3. ECS IAM Roles for Tasks
4. EC2 IMDS (v1 support only for now)
5. EKS IAM Roles for Service Accounts
6. STS creds when you use the `role_arn` option in each output.

If you can add this that'd be awesome, if not I'll add it to my teams backlog.

**Comment by brimworks** [comment]
Unfortunately, the new `aws/flb_aws_credentials_process.c` doesn't compile on alpine linux due to the fact that stdin/stdout/stderr are defined as macros. If the `struct process` was changed so the field names did not conflict with the stdin/stdout/stderr macros, it would make this code easier to compile on other platforms.

Specific error that I see:

```
[ 48%] Building C object src/CMakeFiles/fluent-bit-static.dir/aws/flb_aws_credentials_process.c.o
In file included from /build/lib/monkey/include/monkey/mk_core/mk_iov.h:23,
                 from /build/lib/monkey/include/monkey/mk_core.h:35,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/build/src/aws/flb_aws_credentials_process.c: In function 'new_process':
/build/src/aws/flb_aws_credentials_process.c:436:10: error: expected identifier before '(' token
  436 |         .stdin = -1,
      |          ^~~~~
/build/src/aws/flb_aws_credentials_process.c:437:9: error: expected '}' before '.' token
  437 |         .stdout = {-1, -1},
      |         ^
/build/src/aws/flb_aws_credentials_process.c:433:27: note: to match this '{'
  433 |     *p = (struct process) {
      |                           ^
In file included from /build/lib/monkey/include/monkey/mk_core/mk_iov.h:23,
                 from /build/lib/monkey/include/monkey/mk_core.h:35,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/build/src/aws/flb_aws_credentials_process.c:442:16: error: expected identifier before '(' token
  442 |     while ((p->stdin = open(DEV_NULL, O_RDONLY|O_CLOEXEC)) < 0) {
      |                ^~~~~
/build/src/aws/flb_aws_credentials_process.c:449:28: error: expected identifier before '(' token
  449 |     if (flb_pipe_create(p->stdout) < 0) {;
      |                            ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:454:18: error: expected identifier before '(' token
  454 |     if (fcntl(p->stdout[0], F_SETFL, O_CLOEXEC) < 0) {
      |                  ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:454:9: error: too few arguments to function 'fcntl'
  454 |     if (fcntl(p->stdout[0], F_SETFL, O_CLOEXEC) < 0) {
      |         ^~~~~
In file included from /usr/include/sys/epoll.h:10,
                 from /build/lib/monkey/include/monkey/mk_core/mk_event_epoll.h:20,
                 from /build/lib/monkey/include/monkey/mk_core/mk_event.h:72,
                 from /build/lib/monkey/include/monkey/mk_core.h:58,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/usr/include/fcntl.h:33:5: note: declared here
   33 | int fcntl(int, int, ...);
      |     ^~~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_iov.h:23,
                 from /build/lib/monkey/include/monkey/mk_core.h:35,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/build/src/aws/flb_aws_credentials_process.c:459:18: error: expected identifier before '(' token
  459 |     if (fcntl(p->stdout[1], F_SETFL, O_CLOEXEC) < 0) {
      |                  ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:459:9: error: too few arguments to function 'fcntl'
  459 |     if (fcntl(p->stdout[1], F_SETFL, O_CLOEXEC) < 0) {
      |         ^~~~~
In file included from /usr/include/sys/epoll.h:10,
                 from /build/lib/monkey/include/monkey/mk_core/mk_event_epoll.h:20,
                 from /build/lib/monkey/include/monkey/mk_core/mk_event.h:72,
                 from /build/lib/monkey/include/monkey/mk_core.h:58,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/usr/include/fcntl.h:33:5: note: declared here
   33 | int fcntl(int, int, ...);
      |     ^~~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_iov.h:23,
                 from /build/lib/monkey/include/monkey/mk_core.h:35,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/build/src/aws/flb_aws_credentials_process.c:464:16: error: expected identifier before '(' token
  464 |     while ((p->stderr = open(DEV_NULL, O_WRONLY|O_CLOEXEC)) < 0) {
      |                ^~~~~~
/build/src/aws/flb_aws_credentials_process.c: In function 'exec_process_child':
/build/src/aws/flb_aws_credentials_process.c:482:21: error: expected identifier before '(' token
  482 |     while ((dup2(p->stdin, STDIN_FILENO) < 0)) {
      |                     ^~~~~
/build/src/aws/flb_aws_credentials_process.c:482:13: error: too few arguments to function 'dup2'
  482 |     while ((dup2(p->stdin, STDIN_FILENO) < 0)) {
      |             ^~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_unistd.h:26,
                 from /build/lib/monkey/include/monkey/mk_core.h:39,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/usr/include/unistd.h:40:5: note: declared here
   40 | int dup2(int, int);
      |     ^~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_iov.h:23,
                 from /build/lib/monkey/include/monkey/mk_core.h:35,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/build/src/aws/flb_aws_credentials_process.c:487:21: error: expected identifier before '(' token
  487 |     while ((dup2(p->stdout[1], STDOUT_FILENO) < 0)) {
      |                     ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:487:13: error: too few arguments to function 'dup2'
  487 |     while ((dup2(p->stdout[1], STDOUT_FILENO) < 0)) {
      |             ^~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_unistd.h:26,
                 from /build/lib/monkey/include/monkey/mk_core.h:39,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/usr/include/unistd.h:40:5: note: declared here
   40 | int dup2(int, int);
      |     ^~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_iov.h:23,
                 from /build/lib/monkey/include/monkey/mk_core.h:35,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/build/src/aws/flb_aws_credentials_process.c:492:21: error: expected identifier before '(' token
  492 |     while ((dup2(p->stderr, STDERR_FILENO) < 0)) {
      |                     ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:492:13: error: too few arguments to function 'dup2'
  492 |     while ((dup2(p->stderr, STDERR_FILENO) < 0)) {
      |             ^~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_unistd.h:26,
                 from /build/lib/monkey/include/monkey/mk_core.h:39,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/usr/include/unistd.h:40:5: note: declared here
   40 | int dup2(int, int);
      |     ^~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_iov.h:23,
                 from /build/lib/monkey/include/monkey/mk_core.h:35,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/build/src/aws/flb_aws_credentials_process.c:498:14: error: expected identifier before '(' token
  498 |     close(p->stdin);
      |              ^~~~~
/build/src/aws/flb_aws_credentials_process.c:499:23: error: expected identifier before '(' token
  499 |     flb_pipe_close(p->stdout[0]);
      |                       ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:500:23: error: expected identifier before '(' token
  500 |     flb_pipe_close(p->stdout[1]);
      |                       ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:501:14: error: expected identifier before '(' token
  501 |     close(p->stderr);
      |              ^~~~~~
/build/src/aws/flb_aws_credentials_process.c: In function 'exec_process':
/build/src/aws/flb_aws_credentials_process.c:528:14: error: expected identifier before '(' token
  528 |     close(p->stdin);
      |              ^~~~~
/build/src/aws/flb_aws_credentials_process.c:529:8: error: expected identifier before '(' token
  529 |     p->stdin = -1;
      |        ^~~~~
/build/src/aws/flb_aws_credentials_process.c:531:23: error: expected identifier before '(' token
  531 |     flb_pipe_close(p->stdout[1]);
      |                       ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:532:8: error: expected identifier before '(' token
  532 |     p->stdout[1] = -1;
      |        ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:534:14: error: expected identifier before '(' token
  534 |     close(p->stderr);
      |              ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:535:8: error: expected identifier before '(' token
  535 |     p->stderr = -1;
      |        ^~~~~~
/build/src/aws/flb_aws_credentials_process.c: In function 'read_from_process':
/build/src/aws/flb_aws_credentials_process.c:552:18: error: expected identifier before '(' token
  552 |     if (fcntl(p->stdout[0], F_SETFL, O_NONBLOCK) < 0) {
      |                  ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:552:9: error: too few arguments to function 'fcntl'
  552 |     if (fcntl(p->stdout[0], F_SETFL, O_NONBLOCK) < 0) {
      |         ^~~~~
In file included from /usr/include/sys/epoll.h:10,
                 from /build/lib/monkey/include/monkey/mk_core/mk_event_epoll.h:20,
                 from /build/lib/monkey/include/monkey/mk_core/mk_event.h:72,
                 from /build/lib/monkey/include/monkey/mk_core.h:58,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/usr/include/fcntl.h:33:5: note: declared here
   33 | int fcntl(int, int, ...);
      |     ^~~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_iov.h:23,
                 from /build/lib/monkey/include/monkey/mk_core.h:35,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/build/src/aws/flb_aws_credentials_process.c:570:22: error: expected identifier before '(' token
  570 |             .fd = p->stdout[0],
      |                      ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:620:54: error: expected identifier before '(' token
  620 |             result = read_until_block(p->args[0], p->stdout[0], buf);
      |                                                      ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:620:22: error: too few arguments to function 'read_until_block'
  620 |             result = read_until_block(p->args[0], p->stdout[0], buf);
      |                      ^~~~~~~~~~~~~~~~
/build/src/aws/flb_aws_credentials_process.c:357:12: note: declared here
  357 | static int read_until_block(char* name, flb_pipefd_t fd, struct readbuf* buf)
      |            ^~~~~~~~~~~~~~~~
In file included from /build/lib/monkey/include/monkey/mk_core/mk_iov.h:23,
                 from /build/lib/monkey/include/monkey/mk_core.h:35,
                 from /build/include/fluent-bit/flb_output.h:32,
                 from /build/include/fluent-bit/flb_aws_util.h:24,
                 from /build/include/fluent-bit/flb_aws_credentials.h:25,
                 from /build/src/aws/flb_aws_credentials_process.c:20:
/build/src/aws/flb_aws_credentials_process.c: In function 'destroy_process':
/build/src/aws/flb_aws_credentials_process.c:667:16: error: expected identifier before '(' token
  667 |         if (p->stdin >= 0) {
      |                ^~~~~
/build/src/aws/flb_aws_credentials_process.c:668:22: error: expected identifier before '(' token
  668 |             close(p->stdin);
      |                      ^~~~~
/build/src/aws/flb_aws_credentials_process.c:669:16: error: expected identifier before '(' token
  669 |             p->stdin = -1;
      |                ^~~~~
/build/src/aws/flb_aws_credentials_process.c:671:16: error: expected identifier before '(' token
  671 |         if (p->stdout[0] >= 0) {
      |                ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:672:22: error: expected identifier before '(' token
  672 |             close(p->stdout[0]);
      |                      ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:673:16: error: expected identifier before '(' token
  673 |             p->stdout[0] = -1;
      |                ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:675:16: error: expected identifier before '(' token
  675 |         if (p->stdout[1] >= 0) {
      |                ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:676:22: error: expected identifier before '(' token
  676 |             close(p->stdout[1]);
      |                      ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:677:16: error: expected identifier before '(' token
  677 |             p->stdout[1] = -1;
      |                ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:679:16: error: expected identifier before '(' token
  679 |         if (p->stderr >= 0) {
      |                ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:680:22: error: expected identifier before '(' token
  680 |             close(p->stderr);
      |                      ^~~~~~
/build/src/aws/flb_aws_credentials_process.c:681:16: error: expected identifier before '(' token
  681 |             p->stderr = -1;
      |                ^~~~~~
```

Quick fix was to add this:

```
#ifdef stdin
#undef stdin
#endif
#ifdef stdout
#undef stdout
#endif
#ifdef stderr
#undef stderr
#endif
```

...before the declaration of the `struct process`.

**Comment by PettitWesley** [comment]
@brimworks Please open a PR to fix the build on Alpine and ping me on it.

**Comment by brimworks** [comment]
@PettitWesley I've opened a PR for this here: https://github.com/fluent/fluent-bit/pull/4002

Thanks!

**Pull request merged**